### PR TITLE
Basic EasyMDE Implementation

### PR DESCRIPTION
**Pull request by bmgxyz**
At long last and after great effort, we've found an editor that we like. EasyMDE, which is based on CodeMirror, has nearly all of the features that we want and just generally seems like a good platform on which to build our own editor for `setonotes`.

This PR contains a basic implementation of the EasyMDE editor on top of the existing `setonotes` codebase. It's buggy and incomplete, but the path to a better implementation is clear and short when compared with others. We determined that it would be better to merge an incomplete but functional editor into `develop` now than to wait for a more complete version, since upcoming changes will require some firm decisions about editor implementation.

What follows is a description of the current state of the editor.

## Current Behavior

Upon selecting a note, the user lands on `/page/{id}`, which contains the full-window editor and their note as rendered HTML. This view is called the preview mode. In the preview mode, the user can scroll through their note and interact with the rendered elements as if they were on an ordinary web page (because they mostly are).

At any time, the user may press `i`, which launches the editing mode. The preview is replaced with a decorated Markdown version of the note, and the cursor is automatically moved to the beginning of the note. The user may then navigate the text as they would in any ordinary editor, making their edits at will. When the user makes a change that results in some Markdown syntax, the editor decorates the text appropriately.

If the user wishes to exit the editing mode and return to the preview mode, they may do so by pressing `ESC`. To save their changes, the user may press `CTRL+Enter` in either the editing mode or the preview mode.

## Bugs

- Formatting-related toolbar buttons and keyboard shortcuts are still active in the preview mode.
- When entering the editing mode, the editor scrolls to the top of the note instead of to the scroll position of the preview mode.
- Lines in code blocks are too close together and occlude each other.
- Heading underlines aren't the same width as their words and have too much vertical spacing.
- When the editor is focused, outline color is not the same between browsers.

There are almost certainly more bugs; we just haven't found them yet.

## Future Feature Ideas

- Press `:wq+Enter` in the preview mode to save and exit.
- Press `a` to enter the editing mode (just as `i` does now).
- Use `j` and `k` for scrolling in the preview mode.
- Save without closing (requires API)
- Autosave (also requires API)
- Prompt before closing unsaved
- Code syntax highlighting
- Night mode
- Optional and configurable auto-wrap
- Improve visual design (fonts, text sizing, etc.)
- Allow custom stylesheets
- Help button with modal instructions
- Markdown compliance
    - en dashes / em dashes
    - three dots to ellipsis
    - footnotes
    - tables
    - math
    - option to switch between Markdown flavors (CommonMark, GFM, etc.)
- When returning to the directory, the first note should be focused so the user can use the arrow keys (or `j` and `k`) to move and select another one.
- `Backspace` in the preview mode should go back to the directory.
- Add an actual edit button (which must change to a preview button and back again).
- Enable/disable spellcheck and/or implement a better spellcheck interface
- Add an optional animation between preview and editor mode.
- Double click to launch editor and start editing at the clicked position.

### CODE REVIEW

File: cmd/handlers.go [hunk 1]
```diff
@@ -135,68 +138,12 @@ func newlineDoctor(text []byte) []byte {
...
-    s.renderTemplate(w, "view.tmpl", data)
-}
-
 func (s *server) editHandler(w http.ResponseWriter, r *http.Request, pageID int,
     userID int, authorized bool) {
 
     // redirect visitors
     if !authorized {
-        log.Println("authorized attempt to view /view/")
+        log.Println("unauthorized attempt to edit", pageID)
```
Comment (loboyd) [inline cmd/handlers.go:146]: Nice catch on all of these. Oops!

**Comment by loboyd** [comment]
Nice write-up! I think we should consider *any* focus border on the editor/viewer to be a bug that we should get rid of. I really like most of the feature ideas you have here. I actually thought of the `j` and `k` directory nav the other day too because I really like that Github does that.

Did you put any more thought into the build process now that EasyMDE is included? Is tracking the minified JS files the best thing to do for now?

File: cmd/handlers.go (hunk 1 above)
Comment (bmgxyz) [reply cmd/handlers.go:146]: Thanks. Yeah, I did a double-take on this line initially, but it's a simple fix.

**Comment by bmgxyz** [comment]
> Nice write-up!

Thank you. :rocket:

> I think we should consider _any_ focus border on the editor/viewer to be a bug that we should get rid of.

I agree. That's what I intended to convey with that bug, but I see now that it's not really clear. In any case, it should be a simple task of defining `outline: none` or similar for the editor when it's `focus`-ed.

> Did you put any more thought into the build process now that EasyMDE is included? Is tracking the minified JS files the best thing to do for now?

No, I haven't, but I think arguments can be made for both options (tracking the minified files or not). If we keep them, then deploying `setonotes` becomes a fair bit simpler, but not tracking them is cleaner. As I type this, I'm realizing that tracking the minified files would sort of be like tracking the compiled server binary, which we don't do (and for good reason: it doesn't add any information). We already depend on Go libraries that aren't tracked, so maybe it would be better to leave the minified (or perhaps "compiled") JS and CSS out, too.

I'll push a commit soon that stops tracking them and send another review request to you.

**Comment by loboyd** [comment]
> No, I haven't, but I think arguments can be made for both options (tracking the minified files or not). If we keep them, then deploying setonotes becomes a fair bit simpler, but not tracking them is cleaner. As I type this, I'm realizing that tracking the minified files would sort of be like tracking the compiled server binary, which we don't do (and for good reason: it doesn't add any information). We already depend on Go libraries that aren't tracked, so maybe it would be better to leave the minified (or perhaps "compiled") JS and CSS out, too.

> I'll push a commit soon that stops tracking them and send another review request to you.

Yeah, that's a good analogy. I agree that's the way to go.

**Pull request merged**